hudi-common/src/main/java/org/apache/hudi/common/table/cdc/HoodieCDCUtils.java

```diff
@@ -0,0 +1,150 @@
+  public static final String CDC_AFTER_IMAGE = "after";
+
+  /* the key of the changed record */
+  public static final String CDC_RECORD_KEY = "record_key";
+
+  public static final String[] CDC_COLUMNS = new String[] {
+      CDC_OPERATION_TYPE,
+      CDC_COMMIT_TIMESTAMP,
+      CDC_BEFORE_IMAGE,
+      CDC_AFTER_IMAGE
+  };
+
+  /**
+   * This is the standard CDC output format.
+   * Also, this is the schema of cdc log file in the case `hoodie.table.cdc.supplemental.logging.mode` is 'cdc_data_before_after'.
+   */
+  public static final String CDC_SCHEMA_STRING = "{\"type\":\"record\",\"name\":\"Record\","
+      + "\"fields\":["
+      + "{\"name\":\"op\",\"type\":[\"string\",\"null\"]},"
+      + "{\"name\":\"ts\",\"type\":[\"string\",\"null\"]},"
+      + "{\"name\":\"before\",\"type\":[\"string\",\"null\"]},"
+      + "{\"name\":\"after\",\"type\":[\"string\",\"null\"]}"
+      + "]}";
+
+  public static final Schema CDC_SCHEMA = new Schema.Parser().parse(CDC_SCHEMA_STRING);
+
+  /**
+   * The schema of cdc log file in the case `hoodie.table.cdc.supplemental.logging.mode` is 'cdc_data_before'.
+   */
+  public static final String CDC_SCHEMA_OP_RECORDKEY_BEFORE_STRING = "{\"type\":\"record\",\"name\":\"Record\","
+      + "\"fields\":["
+      + "{\"name\":\"op\",\"type\":[\"string\",\"null\"]},"
+      + "{\"name\":\"record_key\",\"type\":[\"string\",\"null\"]},"
+      + "{\"name\":\"before\",\"type\":[\"string\",\"null\"]}"
+      + "]}";
+
+  public static final Schema CDC_SCHEMA_OP_RECORDKEY_BEFORE =
+      new Schema.Parser().parse(CDC_SCHEMA_OP_RECORDKEY_BEFORE_STRING);
+
+  /**
+   * The schema of cdc log file in the case `hoodie.table.cdc.supplemental.logging.mode` is 'op_key'.
+   */
+  public static final String CDC_SCHEMA_OP_AND_RECORDKEY_STRING = "{\"type\":\"record\",\"name\":\"Record\","
+      + "\"fields\":["
+      + "{\"name\":\"op\",\"type\":[\"string\",\"null\"]},"
+      + "{\"name\":\"record_key\",\"type\":[\"string\",\"null\"]}"
+      + "]}";
+
+  public static final Schema CDC_SCHEMA_OP_AND_RECORDKEY =
+      new Schema.Parser().parse(CDC_SCHEMA_OP_AND_RECORDKEY_STRING);
+
+  public static final Schema schemaBySupplementalLoggingMode(HoodieCDCSupplementalLoggingMode supplementalLoggingMode) {
+    switch (supplementalLoggingMode) {
+      case WITH_BEFORE_AFTER:
+        return CDC_SCHEMA;
+      case WITH_BEFORE:
+        return CDC_SCHEMA_OP_RECORDKEY_BEFORE;
+      case OP_KEY:
+        return CDC_SCHEMA_OP_AND_RECORDKEY;
+      default:
+        throw new HoodieException("not support this supplemental logging mode: " + supplementalLoggingMode);
+    }
+  }
+
+  /**
+   * Build the cdc record which has all the cdc fields when `hoodie.table.cdc.supplemental.logging.mode` is 'cdc_data_before_after'.
+   */
+  public static GenericData.Record cdcRecord(
+      String op, String commitTime, GenericRecord before, GenericRecord after) {
+    String beforeJsonStr = recordToJson(before);
```

Comment:
I think all of my comments remained in the other PR; repeating here:

Why we're using JSON for the records? Shall we use Avro at least to reduce the footprint?

Reply:
BTW, no need to tackle it in this PR, but want to kickoff the conversation on that

Reply:
if the before/after value is store as the json string, we can load the cdc log file and return without any serialize/deserialize. IMO, persisting cdc data is an operation that consumes more storage in exchange for improved query efficiency.

Reply:
i think many users would be concerned about additional storage once enabled cdc with before/after mode. +1 to reduce it. also avro is already adopted for hudi log files, we should at least keep the default format consistent.

Reply:
@YannByron let's make sure we follow-up on this (and let's also create a Jira for it to make sure we're not losing the context)

Reply:
https://issues.apache.org/jira/browse/HUDI-4887
I originally planned to solve it in cdc-reading-related pr, but maybe a separate jira/pr would be more appropriate.